src/lib.rs

```diff
@@ -513,26 +518,29 @@ impl Verification for VerifyOnly {}
 impl Verification for All {}
 
 /// The secp256k1 engine, used to execute all signature operations
-pub struct Secp256k1<C> {
+pub struct Secp256k1<'buf, C> {
     ctx: *mut ffi::Context,
-    phantom: PhantomData<C>
+    phantom: PhantomData<C>,
+    _buf: Option<&'buf [u8]>,
```

Comment:
I think unconditionally adding `'buf` to the public API will make things too unergonomic for users who aren't providing their own buffer.

I think it'd be better to add a new set of context types `SignOnlyBuffer<'buf>` `VerifyOnlyBuffer<'buf>` `AllBuffer<'buf>` which tie the lifetime of the passed-in buffer to the context. Then let `phantom` deal with the borrowck and change the `_buf` field into `buf: Option<*mut u8>`.

Reply:
hmm you prefer that `no-std` users will have a different Context struct? (If you provide no buffer you shouldn't feel the new lifetime at all(as a user, for the library maintainers you're right that it's unergonomic))

Reply:
If we do go in that direction we'll need to abstract out the Context, so that sign/verify functions won't be implemented on the struct itself but on a generic of the trait (adding something like `get_ptr()` to the traits)

Reply:
the nice thing about having `&[u8]` is that it's a fat pointer which makes it easier to clone (no need to recheck the size)

Reply:
At the very least having `'buf` in the API makes the documentation confusing, since users will see a lifetime but have no idea what it's for or how to use it.

I'm pretty sure sign/verify are already abstracted over the context, since e.g. both `SignOnly` and `All` have signing functionality.